Storage/SQL/SQLiteHistory.swift

```diff
@@ -915,49 +915,66 @@ extension SQLiteHistory: SyncableHistory {
 
         // Turn our lazy collection of integers into a comma-seperated string for the IN clause.
         let historyIDs = Array(places.keys)
-        let inClause = "siteID IN ( \(historyIDs.map(String.init).joined(separator: ",")) )"
 
         let sql =
-        "SELECT v1.siteID AS siteID, v1.date AS visitDate, v1.type AS visitType " +
-        "FROM (" +
-        "   SELECT * FROM \(TableVisits) WHERE \(inClause) AND type <> 0" +
-        ") AS v1 " +
-        "LEFT OUTER JOIN \(TableVisits) AS v2 ON v1.siteID = v2.siteID AND v1.date < v2.date " +
-        "GROUP BY v1.date " +
-        "HAVING COUNT(*) < ? " +
-        "ORDER BY v1.siteID, v1.date DESC"
-
-        // Seed our accumulator with empty lists since we already know which IDs we will be fetching.
-        var visits = [Int: [Visit]]()
-        historyIDs.forEach { visits[$0] = [] }
-
-        // Add each visit to its history item's list.
-        let visitsAccumulator: (SDRow) -> Void = { row in
-            let date = row.getTimestamp("visitDate")!
-            let type = VisitType(rawValue: row["visitType"] as! Int)!
-            let visit = Visit(date: date, type: type)
-            let id = row["siteID"] as! Int
-            visits[id]?.append(visit)
-        }
+                "SELECT siteID, date AS visitDate, type AS visitType " +
+                "FROM \(TableVisits) " +
+                "WHERE siteID IN (\(historyIDs.map(String.init).joined(separator: ","))) " +
+                "ORDER BY date DESC"
+
+        // We want to get a tuple Visit and Place here. We can either have an explicit tuple factory
+        // or we use an identity function, and make do without creating extra data structures.
+        // Since we have a known Out Of Memory issue here, let's avoid extra data structures.
+        let rowIdentity: (SDRow) -> SDRow = { $0 }
+
+        // We'll need to runQueryUnsafe so we get a LiveCursor, i.e. we don't get the cursor
+        // contents into memory all at once.
+        return db.runQueryUnsafe(sql, args: nil, factory: rowIdentity) >>== { cursor in
+            defer { cursor.close() }
+
+            // Seed our accumulator with empty lists since we already know which IDs we will be fetching.
+            var visits = [Int: [Visit]]()
+            historyIDs.forEach { visits[$0] = [] }
+
+            // We need to iterate through these explicitly, without relying on the
+            // factory.
+            for row in cursor.makeIterator() {
+                guard let row = row else { break }
+                let date = row.getTimestamp("visitDate")!
+                let type = VisitType(rawValue: row["visitType"] as! Int)!
+                let visit = Visit(date: date, type: type)
+                let id = row["siteID"] as! Int
+
+                // Append the visits in descending date order, so we only get the
+                // most recent top N.
+                if visits[id]?.count ?? visitLimit < visitLimit {
```

Comment:
You should do this work much earlier in the block — if you have more than 20 visits for this site, you want to skip through the cursor to the next site ASAP, rather than building a `Visit` only to drop it on the floor.

Reply:
Also I think this conditional is wrong, no? When `visits` is empty, `visits[123]?.count ?? visitLimit` will be `visitLimit`, and `visitLimit < visitLimit` is false, so we will never record any visits!

Reply:
Oh, we're pre-seeding it. Yeah, that'll make this OK.

Reply:
It's confusing enough to make it worth re-writing though.

It disappears when you put the check in a `guard` and move it to the top of the loop.